aten/src/ATen/native/cudnn/RNN.cpp

```diff
@@ -1007,6 +1049,234 @@ Tensor _cudnn_init_dropout_state(const Type& ty, double dropout, bool train, int
   return dropout_desc.state;
 }
 
+////////////////////////////////////////////////////////////////////////////////
+// CUDA dispatch for the generic RNN ops (at::lstm, at::gru, ...)
+////////////////////////////////////////////////////////////////////////////////
+
+namespace {
+
+// Helpers for working with different hidden types.
+std::tuple<Tensor, Tensor> unpack_hidden(const Tensor& hidden) {
+  return std::make_tuple(hidden, at::Tensor{});
+}
+
+std::tuple<Tensor, Tensor> unpack_hidden(const std::tuple<Tensor, Tensor>& hidden) {
+  return hidden;
+}
+
+template<typename hidden_type>
+hidden_type pack_hidden(const Tensor& hx, const Tensor& cx) {
+  static_assert(std::is_same<hidden_type, void>::value, "pack_hidden not implemented for this type");
+  AT_ERROR("NOT IMPLEMENTED");
+}
+
+template<>
+Tensor pack_hidden<Tensor>(const Tensor& hx, const Tensor& cx) {
+  AT_ASSERT(cx.numel() == 0);
+  return hx;
+}
+
+template<>
+std::tuple<Tensor, Tensor> pack_hidden<std::tuple<Tensor, Tensor>>(const Tensor& hx, const Tensor& cx) {
+  return std::make_tuple(hx, cx);
+}
+
+struct DropoutState {
```

Comment:
Can you comment on the event structure used in dropout state. I think I understand what is going on but it took some time to unwind.

Reply:
Sure. It's really simple - every time you start using a dropout state, you synchronize with its event (to make sure all previous uses finish before this one), and once you're done you record the event to allow others to synchronize with your kernel. Those events are really needed only for inter-stream sync on a single GPU. I doubt anyone will want to run cuDNN RNNs in parallel on a single GPU, so they should end up being complete no-ops most of the time.

Reply:
I'll land this so that I can send the next PR, and I'll add a comment in another PR.